AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift

```diff
@@ -0,0 +1,144 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+// swiftlint:disable cyclomatic_complexity
+// swiftlint:disable syntactic_sugar
+extension ModelSchema {
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    public static func isEqual(_ model1: Model,
+                               _ model2: Model,
+                               _ modelSchema: ModelSchema) -> Bool {
+        for (fieldName, modelField) in modelSchema.fields {
+            // skip read only fields
+            if modelField.isReadOnly {
+                continue
+            }
+
+            let value1 = model1[fieldName] ?? nil
+            let value2 = model2[fieldName] ?? nil
+
+            // check equality for different `ModelFieldType`
+            switch modelField.type {
+            case .string:
+                guard let value1Optional = value1 as? String?, let value2Optional = value2 as? String? else {
+                    return false
+                }
```

Comment:
what if we have a `modelField.type` and then `value1` and `value2` are both `nil`- meaning both values do not exist on either models. it should continue because there is nothing to check but this returns false and says the model's aren't equal